okhttp/src/main/java/okhttp3/internal/PublicSuffixDatabase.java

```diff
@@ -0,0 +1,346 @@
+    StringBuilder effectiveTldPlusOne = new StringBuilder();
+    for (int i = labelsToInclude; i < domainToReturnLabels.length; i++) {
+      effectiveTldPlusOne.append(domainToReturnLabels[i]).append('.');
+    }
+    effectiveTldPlusOne.deleteCharAt(effectiveTldPlusOne.length() - 1);
+
+    return effectiveTldPlusOne.toString();
+  }
+
+  private String[] findMatchingRule(String[] domainLabels) {
+    if (!listRead.get() && listRead.compareAndSet(false, true)) {
+      readTheList();
+    } else {
+      try {
+        readCompleteLatch.await();
+      } catch (InterruptedException ignored) {
+      }
+    }
+
+    synchronized (this) {
+      if (publicSuffixListBytes == null) {
+        // We failed to read the public suffix list. This scenario should never happen, but in case
+        // it does, we'll let act as though this is a no-op.
+        return null;
+      }
+    }
+
+    // Break apart the domain into UTF-8 labels, i.e. foo.bar.com turns into [foo, bar, com].
+    byte[][] domainLabelsUtf8Bytes = new byte[domainLabels.length][];
+    for (int i = 0; i < domainLabels.length; i++) {
+      domainLabelsUtf8Bytes[i] = domainLabels[i].getBytes(Util.UTF_8);
+    }
+
+    // Start by looking for exact matches. We start at the leftmost label. For example, foo.bar.com
+    // will look like: [foo, bar, com], [bar, com], [com]. The longest matching rule wins.
+    String exactMatch = null;
+    for (int i = 0; i < domainLabelsUtf8Bytes.length; i++) {
+      String match = binarySearchBytes(publicSuffixListBytes, domainLabelsUtf8Bytes, i);
+      if (match != null) {
+        exactMatch = match;
+        break;
+      }
+    }
+
+    // Wildcard rules throw a wrench into the mix. They are not restricted to appear in the leftmost
+    // position. An example is: *.*.foo. Therefore, we must exhaust the permutations,
+    // substituting a wildcard in the various label positions. For example, on the domain
+    // foo.bar.com, we'll try: [*, com], [foo, *, com], [*, bar, com], [*, *, com].
+    // Note: since we've already tried an exact match, we leave out those permutations.
+    String wildcardMatch = null;
+    if (domainLabelsUtf8Bytes.length > 1) {
+      // It seems highly unlikely (impossible?) the rightmost label will ever be a wildcard.
+      byte[][] labelsWithWildcard = new byte[domainLabelsUtf8Bytes.length][];
+      labelsWithWildcard[domainLabelsUtf8Bytes.length - 1] =
+          domainLabelsUtf8Bytes[domainLabelsUtf8Bytes.length - 1];
+
+      int firstLevel = domainLabelsUtf8Bytes.length - 1;
+      outer: for (int currentLevel = 0; currentLevel < firstLevel; currentLevel++) {
+        int numberOfWildcards = firstLevel - currentLevel;
+        int permutations = 1 << numberOfWildcards;
+        for (int i = 1; i < permutations; i++) {
+          for (int labelPos = firstLevel - 1; labelPos >= currentLevel; labelPos--) {
+            // Use the bit positions to indicate whether we should place a wildcard there. For
```

Comment:
I wrestled with this a bit more. I backed out of the complicated implementation. The specification as defined is very flexible. Wildcards are not restricted to the leftmost position. In practice though, they are all in the leftmost position. In effect, the code is testing permutations that do not exist and essentially doing unnecessary work.

The code cheats now, and only considers the leftmost position. I've incorporated some assertions in the generator to confirm this. It will mean if the very flexible specification is ever exercised, the implementation will need to be revisited.

Let me know if you think this is a bad decision.